File: DataEngineering/06_Data_Orchestration/projects/python/02_task_retry_handler/main.py
import random
import logging
from tenacity import retry, wait_exponential, stop_after_attempt, retry_if_exception_type, before_sleep_log

# Configure logging
logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)

class FlakyTaskError(Exception):
    pass

# Retry configuration:
# - Exponential backoff (wait = 2^x)
# - Max wait time of 10 seconds between retries
# - Stop after 5 total attempts
# - Only retry on FlakyTaskError
@retry(
    wait=wait_exponential(multiplier=1, min=1, max=10),
    stop=stop_after_attempt(5),
    retry=retry_if_exception_type(FlakyTaskError),
    before_sleep=before_sleep_log(logger, logging.WARNING)
)
def flaky_data_pull():
    logger.info("Attempting to pull data from upstream API...")
    if random.random() < 0.7:
        raise FlakyTaskError("Upstream API returned 503 Service Unavailable")
    
    logger.info("Data pull succeeded!")
    return {"data": [1, 2, 3]}

if __name__ == "__main__":
    try:
        result = flaky_data_pull()
        logger.info(f"Final result: {result}")
    except Exception as e:
        logger.error(f"Task ultimately failed after retries: {e}")

File: DataEngineering/06_Data_Orchestration/projects/golang/02_task_retry_handler/main.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"time"

	"github.com/avast/retry-go"
)

// flakyDataPull simulates pulling data from an unstable API.
func flakyDataPull() (string, error) {
	log.Println("Attempting to pull data from upstream API...")
	if rand.Float32() < 0.7 {
		return "", fmt.Errorf("upstream API returned 503 Service Unavailable")
	}
	return "data_payload", nil
}

func main() {
	rand.Seed(time.Now().UnixNano())

	var result string

	// Using avast/retry-go to handle retries with exponential backoff
	err := retry.Do(
		func() error {
			var err error
			result, err = flakyDataPull()
			return err
		},
		retry.Attempts(5),
		retry.Delay(1*time.Second),
		retry.MaxDelay(10*time.Second),
		retry.DelayType(retry.BackOffDelay),
		retry.OnRetry(func(n uint, err error) {
			log.Printf("Retry %d: %v\n", n, err, n)
		}),
	)

	if err != nil {
		log.Fatalf("Task ultimately failed after retries: %v\n", err)
	}

	log.Printf("Task succeeded! Final result: %s\n", result)
}
